### PR TITLE
Use profile specified in --profile with dbt init

### LINKED ISSUE #6154: [CT-1418] [Bug] CLI argument `--profile` ignored in `dbt init`
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

According to `dbt init --help`, the `--profile` option lets the user load an existing profile to use with a new project. However, that option currently does nothing when you call it with `dbt init`. 

Example:
```
dbt init --profile default proj1
```

Result:
proj1 created with `proj1/dbt_project.yml` showing `profile: 'proj1'`. `~/.dbt/profiles.yml` shows a new profile `proj1` created with placeholders. **Note**: if a profile named `proj1` already exists, dbt *still* creates another profile named `proj1` again.

In contrast, calling `dbt run --profile default` actually does make the project use the profile `default`, so the `--profile` argument does work elsewhere at least. 

### Expected Behavior

proj1 created with `proj1/dbt_project.yml` showing `profile: 'default'`. `~/.dbt/profiles.yml` should not show any new profile created.

### Steps To Reproduce

1. Create new project directory to store project in
2. Open `~/.dbt/profiles.yml` and create a new profile `default`.
3. Run `dbt init --profile default proj1`
4. `cd` into `/proj1` and open `dbt_project.yml`. The `profile` will be set to `proj1` instead of `default`.
5. Check `~/.dbt/profiles.yml` again and see that dbt has created a new blank `proj1` profile to go with the new project, instead of using the existing `default` profile.

### Relevant log output

_No response_

### Environment

```markdown
- OS: Ubuntu 22.04.1 LTS
- Python: 3.10.6
- dbt: 1.3.0
```

### Which database adapter are you using with dbt?

postgres

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by ezraerb**
resolves #6154

### Description

If --profile is specified with dbt init, and the profile exists, initialize the project with that as the profile instead of creating a new profile. If the profile does not exist, or the command is run inside an existing project, it fails with an error.

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [X] I have run this code in development and it appears to resolve the stated issue
- [X] This PR includes tests, or tests are not required/relevant for this PR
- [X] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [X] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Comment by dbeatty10** [comment]
Thanks for submitting this pull request @ezraerb !

The core team is hard at work making lots of improvements so it may be a while before an engineer is assigned to work on this PR.

In the meantime, https://github.com/dbt-labs/dbt-core/pull/7609 is updating the same area of code (to resolve a different issue), so one or the other will need to be updated depending on the merge order. We'll probably want to merge https://github.com/dbt-labs/dbt-core/pull/7609 first, so you'd need to resolve the conflicts afterwards.

File: .changes/unreleased/Fixes-20230424-161642.yaml
Comment (dbeatty10) [inline .changes/unreleased/Fixes-20230424-161642.yaml:7]: For `changie`, we use the GitHub issue number rather than the Jira issue number.
```suggestion
  Issue: "6154"
```

**Comment by ezraerb** [comment]
Sure thing. Let me know when I need to make edits after the other merge.

**Comment by dbeatty10** [comment]
@ezraerb https://github.com/dbt-labs/dbt-core/pull/7609 is merged now ✅

**Comment by ezraerb** [comment]
Merged code has been pushed to my branch

File: core/dbt/task/init.py
Comment (VersusFacit) [inline core/dbt/task/init.py:282]: Let's import and use an appropriate exception from `dbt.exceptions` which is defined in `core/dbt/exceptions`. Maybe `DbtRuntimeError`?

File: core/dbt/task/init.py
Comment (VersusFacit) [inline core/dbt/task/init.py:306]: ditto for exceptions

File: tests/functional/init/test_init.py
Comment (VersusFacit) [inline tests/functional/init/test_init.py:90]: we have a builtin flag for this

`run_dbt(..., expect_pass=False)`

File: tests/functional/init/test_init.py
Comment (VersusFacit) [inline tests/functional/init/test_init.py:180]: ditto

File: tests/functional/init/test_init.py
Comment (VersusFacit) [inline tests/functional/init/test_init.py:763]: Can we make this a fixture in a constant at file top or in a `fixture.py`?

File: tests/functional/init/test_init.py
Comment (VersusFacit) [inline tests/functional/init/test_init.py:818]: another refactor opportunity

File: tests/functional/init/test_init.py
Comment (VersusFacit) [inline tests/functional/init/test_init.py:851]: ditto

File: core/dbt/task/init.py
Comment (VersusFacit) [inline core/dbt/task/init.py:290]: Sentinel values like this in if-else branches can make it hard to read code and track execution semantics. Would much prefer a different strategy like calling functions that make sense in each if-else branch.

File: tests/functional/init/test_init.py
Comment (VersusFacit) [inline tests/functional/init/test_init.py:743]: Curious why Black didn't reformat this or the line below it to be the same.

File: core/dbt/task/init.py
Comment (VersusFacit) [inline core/dbt/task/init.py:281]: `self.args` will give you access to the profile flag's value, so use that instead of `getattr(get_flags(), "PROFILE", None)`, as this is how we do it elsewhere in core tasks

**Review by VersusFacit — COMMENTED** [review]
Hey there! Thanks a bundle for all this great work to pull these arguments through such that they obey the stated flag interface!! 🙇‍♀️ 

I had some time for reviewing PRs and do a bunch of `task` directory work, so I offered some thought for how to rework your code to be inline with our conventions and practices.

I may not be the one to pull this PR over the finish line, but I hope my comments help you on your way. Welcome to the community as a first-time contributor 🎉

File: core/dbt/task/init.py
Comment (dbeatty10) [inline core/dbt/task/init.py:317]: ### User experience
From a user experience perspective, I don't think we need to raise an error if the specified profile is not found. I'd rather just create it anytime it doesn't exist.

### Implementation
I defer to whoever ends up being the code reviewer for this, but see below for some suggestions for refactoring.

The `run` method is long and has a lot of conditionals, which makes it harder to read. Refactoring this would make it easier to maintain in the future.

So I'd suggest refactoring this logic into its own method (similar to how `check_if_can_write_profile` is its own method). Maybe something similar this (completely untested!):

```python
def check_if_profile_exists(self, profile_name: Optional[str] = None) -> bool:
    profile_exists = False  # assume it doesn't exist unless proven otherwise
    user_profile_name = getattr(get_flags(), "PROFILE", None)
    profiles_dir = get_flags().PROFILES_DIR
    if user_profile_name:
        raw_profiles = read_profile(profiles_dir)
        profile_exists = user_profile_name in raw_profiles
    return profile_exists
```

Then this method can be applied in one or more places to use the specified profile if it exists (and create it otherwise).

File: core/dbt/task/init.py
Comment (ezraerb) [reply core/dbt/task/init.py:306]: ditto for the changes.

File: tests/functional/init/test_init.py
Comment (ezraerb) [reply tests/functional/init/test_init.py:90]: Checking the code for run_dbt() shows that expect_pass only applies to the reported status of the task, not any generated exceptions. Those get reraised.

File: tests/functional/init/test_init.py
Comment (ezraerb) [reply tests/functional/init/test_init.py:180]: ditto

File: tests/functional/init/test_init.py
Comment (ezraerb) [reply tests/functional/init/test_init.py:818]: Changed along the lines of the the other tests referenced above.

File: tests/functional/init/test_init.py
Comment (ezraerb) [reply tests/functional/init/test_init.py:851]: ditto

File: tests/functional/init/test_init.py
Comment (ezraerb) [reply tests/functional/init/test_init.py:763]: Checking the other init tests shows that all of them use this pattern. Its probably worth getting more opinions before changing it.

File: core/dbt/task/init.py
Comment (ezraerb) [reply core/dbt/task/init.py:317]: For user experience, I think more input is needed. The risk with creating the profile if it does not exist is the classic typo problem, where a misspelling creates a new profile instead of using the one the user actually wanted. The requirements said "existing profile" so I put in the check.

For implementation, the test is only done once currently, but shrinking a big method is usually a good idea. Refactored.

Comment (dbeatty10) [reply core/dbt/task/init.py:317]: Did you push the refactored code @ezraerb?

This is the most recent that I'm seeing:
image

Comment (ezraerb) [reply core/dbt/task/init.py:317]: Still working on a few other comments, so have not pushed yet.

File: core/dbt/task/init.py
Comment (ezraerb) [reply core/dbt/task/init.py:290]: I personally can handle sentinel values as long as the names are descriptive of the state they capture, and use is not excessive. However, there was a request elsewhere to refactor run() into smaller methods. Since creating the profile is triggered in two places, its an excellent candidate for a refactor. Changed as requested.

File: core/dbt/task/init.py
Comment (ezraerb) [reply core/dbt/task/init.py]: Changes have been pushed.

File: core/dbt/task/init.py
Comment (dbeatty10) [inline core/dbt/task/init.py:318]: @ezraerb Can you help me understand why we'd want or need `self.get_profile_name_from_current_project()` to be guarded by `if not self.args.skip_profile_setup:`?

Could we safely do the following outside of `if` statement?
```python
            profile_name = self.get_profile_name_from_current_project()
```

If so, then we could avoid repeating `self.setup_profile(profile_name)` twice (once for each case) and instead we can just do it once at the very end (IFF profile setup is being skipped).

Comment (ezraerb) [reply core/dbt/task/init.py:318]: There are four cases that need to be considered:
1. Initializing inside an existing project, no profile specified. The profile is created (the code being commented on)
2. Initializing inside an existing project, profile specified. This raises an error.
3. Initializing outside an existing project, no profile specified. Create the project and then create the profile.
4. Initializing outside an existing project, profile specified. Create the project only.

The difference between cases 3 and 4 prohibits having a catch-all profile creation at the end of the method. Either signal variables are needed or the calls need to be duplicated in the various paths. Previous viewers expressed a desire for the latter as they believe it is easier to follow.

File: core/dbt/task/init.py
Comment (jtcohen6) [reply core/dbt/task/init.py]: My first instinct was to agree with @dbeatty10 on this point:

> From a user experience perspective, I don't think we need to raise an error if the specified profile is not found. I'd rather just create it anytime it doesn't exist.

I definitely appreciate the typo annoyance: `dbt init --profile defaut` would lead to the writing of a whole new `defaut` profile, when you _intended_ to use the existing `default` profile. In that case, it would be better to get an explicit error.

But it means that, when initializing a new project from scratch, you have exactly two options:
1. Do not pass `--profile` flag: Initialize a new project and a new profile, with the same name as your project.
2. Pass the `--profile` flag. It must match an existing profile.

If we took Doug's recommendation, there would be three options:
1. Do not pass `--profile` flag to initialize a new project and a new profile. The profile name will match your supplied project name (reasonable default behavior).
2. Pass the `--profile` flag, and its value matches an existing profile: Initialize a new project, do not write a new profile.
3. Pass the `--profile` flag, and its value doesn't match an existing profile: Initialize a new project and a new profile. The profile name will match what you passed into the `--profile` flag.

We could even take this one step further, and provide [the same flexibility that we offer when running `dbt init` within an existing project](https://github.com/dbt-labs/dbt-core/blob/b39eeb328c9a245f21b54fb48f2761ad03d4d6ad/core/dbt/task/init.py#L202-L205):
```
The profile <profile-name> already exists in /Users/jerco/.dbt/profiles.yml. Do you wish to overwrite it? [y/N]: N
```

---

I don't have strong feelings either way. As a heuristic to make this determination, I'm thinking about how there are two "modes" of `dbt init`:
1. **Interactive.** Likely first time using dbt. Need to set up everything.
2. **Programmatic.** Someone who has used dbt before, likely on this machine. Wants to skip the click interactivity and jump straight to `dbt init <project_name> --skip-profile-setup`, or `dbt init <project_name> --profile <existing_profile_name>`.

This `--profile` flag feels designed for persona / use case (2). The first user is less likely to want fine-grained control over exactly how the profile is being named — we should provide the easiest path from start to finish, with some sensible defaults along the way. And second user (slightly more experienced) is less likely to want the interactive walkthrough for setting up their profile.

Which is to say: I'm convinced enough that the behavior implemented in this PR is an acceptable user experience. We'll need to document the behavior in a new "Existing profile" section here: https://docs.getdbt.com/reference/commands/init

**Review by emmyoop — CHANGES_REQUESTED** [review]
@ezraerb this is looking great!  We have some time set aside to help you get this over the finish line an merged in.

There's just one more small change I'd like to request.  Updating the help text will make it more obvious that the `--profile` flag must refer to an existing profile.

The help text is [here](https://github.com/dbt-labs/dbt-core/blob/b39eeb328c9a245f21b54fb48f2761ad03d4d6ad/core/dbt/cli/params.py#L318).  Just adding in the word "existing" would be beneficial:

```
help="Which existing profile to load. Overrides setting in dbt_project.yml.",
```

Other than that you'll need to rebase off `main` to pull in the most recent version of core so we can run some tests.  I'm not seeing a reason the current test is failing but we can get a better picture once you branch is updated.

**Comment by ezraerb** [comment]
I pushed an update with the help text update. When I did a merge from the main branch from upstream, git claimed there was nothing to merge. Should I do a full rebase instead?

**Review by emmyoop — APPROVED** [review]
@ezraerb this looks great!  Thanks for making that last small change.  The tests are passing and we're ready to merge this in!  Thanks for your work on getting this fixed!

**Pull request merged**
